### PR TITLE
refactor: add EIP-155 to transaction types

### PR DESCRIPTION
**Pull request by Wodann**

### CODE REVIEW

File: crates/rethnet_eth/src/remote/eth.rs [hunk 1]
```diff
@@ -172,19 +172,37 @@ impl TryFrom<Transaction> for (SignedTransaction, Address) {
...
-                value: value.value,
-                input: value.input,
-                signature: Signature {
-                    r: value.r,
-                    s: value.s,
-                    v: value.v,
-                },
-            }),
+            0 => {
+                if value.v > 36 {
```
Comment (agostbiro) [inline crates/rethnet_eth/src/remote/eth.rs:176]: I saw this check in two places, maybe we could add a function/method for deciding whether `v` is an EIP-155 value?

File: crates/rethnet_eth/src/transaction/signed/eip155.rs [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    /// Recovers the Ethereum address which was used to sign the transaction.
+    pub fn recover(&self) -> Result<Address, SignatureError> {
+        self.signature
+            .recover(EIP155TransactionRequest::from(self).hash())
+    }
+
+    pub fn chain_id(&self) -> u64 {
+        (self.signature.v - 35) / 2
```
Comment (agostbiro) [inline crates/rethnet_eth/src/transaction/signed/eip155.rs:37]: Is this correct if the parity is odd? Would be good to have some tests here

File: crates/rethnet_eth/src/transaction/signed/eip155.rs [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    }
+
+    /// See <https://github.com/ethereum/EIPs/blob/master/EIPS/eip-155.md>
+    /// > If you do, then the v of the signature MUST be set to {0,1} + `CHAIN_ID` * 2 + 35 where
+    /// > {0,1} is the parity of the y value of the curve point for which r is the x-value in the
+    /// > secp256k1 signing process.
+    pub fn meets_eip155(&self, chain_id: u64) -> bool {
```
Comment (agostbiro) [inline crates/rethnet_eth/src/transaction/signed/eip155.rs:44]: Should we reject `chain_id == 0`? Would be good to have some tests here as well for edge cases

File: crates/rethnet_evm_napi/src/block.rs [hunk 4]
```diff
@@ -308,7 +308,10 @@ impl Block {
             .iter()
             .map(|transaction| match transaction {
                 rethnet_eth::transaction::SignedTransaction::Legacy(transaction) => {
-                    LegacySignedTransaction::new(&env, transaction).map(Either3::A)
+                    LegacySignedTransaction::from_legacy(&env, transaction).map(Either3::A)
+                }
+                rethnet_eth::transaction::SignedTransaction::EIP155(transaction) => {
+                    LegacySignedTransaction::from_eip155(&env, transaction).map(Either3::A)
```
Comment (agostbiro) [inline crates/rethnet_evm_napi/src/block.rs:314]: I found this a bit confusing. Are we returning a `LegacySignedTransaction` here on purpose?

File: crates/rethnet_evm_napi/src/transaction/pending.rs [hunk 5]
```diff
@@ -82,7 +82,10 @@ impl PendingTransaction {
     > {
         match &*self.inner {
             rethnet_eth::transaction::SignedTransaction::Legacy(transaction) => {
-                LegacySignedTransaction::new(&env, transaction).map(Either3::A)
+                LegacySignedTransaction::from_legacy(&env, transaction).map(Either3::A)
+            }
+            rethnet_eth::transaction::SignedTransaction::EIP155(transaction) => {
+                LegacySignedTransaction::from_eip155(&env, transaction).map(Either3::A)
```
Comment (agostbiro) [inline crates/rethnet_evm_napi/src/transaction/pending.rs:88]: I'm also confused here if returning `LegacySignedTransaction` is correct

**Review by agostbiro — CHANGES_REQUESTED** [review]
Thanks for the PR @Wodann ! I left some comments

File: crates/rethnet_evm_napi/src/block.rs (hunk 4 above)
Comment (Wodann) [reply crates/rethnet_evm_napi/src/block.rs:314]: Yes, javascript doesn't have strongly typed enums. Instead Typescript determines which variant something is based on the presence of variant fields. Given that both the EIP-155 and Legacy versions have the same fields, but different interpretations, we cannot express this in Typescript/Javascript.

I originally did, but it introduced a bug where both Legacy and EIP-155 signed transactions were interpreted as the former (or rather, the first in the list of enum variants that matches all fields)

File: crates/rethnet_evm_napi/src/transaction/pending.rs (hunk 5 above)
Comment (Wodann) [reply crates/rethnet_evm_napi/src/transaction/pending.rs:88]: https://github.com/NomicFoundation/hardhat/pull/4298#discussion_r1295003527

File: crates/rethnet_eth/src/transaction/signed/eip155.rs (hunk 2 above)
Comment (Wodann) [reply crates/rethnet_eth/src/transaction/signed/eip155.rs:37]: I'm happy to add tests, but would you mind if I do that in https://github.com/NomicFoundation/hardhat/pull/4300/files/3c85628c4f99db9b9274efecce0558d937756f5b#diff-6ba4eabcc2f6a81547196c40a9e9a9ecdc9a53d1a3b6a75012953e6f1960fe0b, since that adds a bunch of tests and would cause conflicts if I also add a module here 😅

File: crates/rethnet_eth/src/remote/eth.rs (hunk 1 above)
Comment (Wodann) [reply crates/rethnet_eth/src/remote/eth.rs]: Addressed in https://github.com/NomicFoundation/hardhat/pull/4298/commits/99653bcb79aee565cae47c479caa4a7c03121da9

File: crates/rethnet_eth/src/transaction/signed/eip155.rs (hunk 3 above)
Comment (Wodann) [reply crates/rethnet_eth/src/transaction/signed/eip155.rs]: I removed these functions in https://github.com/NomicFoundation/hardhat/pull/4298/commits/99653bcb79aee565cae47c479caa4a7c03121da9, as they are no longer necessary.

File: crates/rethnet_eth/src/transaction/signed/eip155.rs (hunk 2 above)
Comment (agostbiro) [reply crates/rethnet_eth/src/transaction/signed/eip155.rs:37]: Sure, sounds good!

**Comment by Wodann** [comment]
Closed in favour of a combined PR: https://github.com/NomicFoundation/hardhat/pull/4300
